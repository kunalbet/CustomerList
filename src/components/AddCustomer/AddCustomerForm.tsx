import { Field, Formik, FormikHelpers } from "formik";
import * as React from "react";
import { ICustomer, Customer } from "../../types/types";
import {
  StyledForm,
  StyledInput,
  StyledLabel,
  StyledAddButton,
} from "./StyledAddCustomerForm";

type Props = {
  saveCustomer: (customer: ICustomer | any) => void;
};

export const AddCustomerForm: React.FC<Props> = ({ saveCustomer }) => {
  return (
    <Formik
      initialValues={{
        firstName: "",
        lastName: "",
        phoneNumber: "",
      }}
      onSubmit={(
        values: Customer,
        { setSubmitting }: FormikHelpers<Customer>
      ) => {
        saveCustomer(values);
        setSubmitting(false);
      }}
    >
      <StyledForm>
        <StyledLabel htmlFor="firstName">First Name</StyledLabel>
        <Field
          as={StyledInput}
          id="firstName"
          name="firstName"
          placeholder="John"
        />

        <StyledLabel htmlFor="lastName">Last Name</StyledLabel>
        <Field
          as={StyledInput}
          id="lastName"
          name="lastName"
          placeholder="Doe"
        />

        <StyledLabel htmlFor="phoneNumber">Phone Number</StyledLabel>
        <Field
          as={StyledInput}
          id="phoneNumber"
          name="phoneNumber"
          placeholder="john@acme.com"
          type="tel"
        />

        <StyledAddButton type="submit">Add Customer</StyledAddButton>
      </StyledForm>
    </Formik>
  );
};
