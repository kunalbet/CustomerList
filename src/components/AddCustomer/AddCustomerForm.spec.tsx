import * as React from "react";
import { render, waitFor, fireEvent, screen } from "../../utils/testUtils";
import { AddCustomerForm } from "./AddCustomerForm";
import userEvent from '@testing-library/user-event';

describe("<AddCustomerForm />", () => {
  it("should render a <AddCustomerForm />", () => {
    const wrapper = render(<AddCustomerForm saveCustomer={() => {}} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('It should add new customer and invoke the saveCustomer prop', async () => {
    const mockSaveCustomer = jest.fn();
    let wrapper = render(<AddCustomerForm saveCustomer={mockSaveCustomer} />);
    const inputFirstName = wrapper.getByLabelText(/First Name/);
    const inputLastName = wrapper.getByLabelText(/Last Name/);
    const inputPhone = wrapper.getByLabelText(/Phone Number/);
    userEvent.type(inputFirstName, 'Tom');
    userEvent.type(inputLastName, 'Smith');
    userEvent.type(inputPhone, '0444333222');
    fireEvent.click(screen.getByTestId('addCust-button'));

    await waitFor(() => {
      expect(inputFirstName).toHaveValue('Tom');
      expect(inputLastName).toHaveValue('Smith');
      expect(inputPhone).toHaveValue('0444333222');
      expect(mockSaveCustomer).toHaveBeenCalledTimes(1);
    });
  });
  
});
