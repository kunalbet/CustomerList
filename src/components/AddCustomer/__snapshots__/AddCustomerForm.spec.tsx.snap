// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCustomerForm /> should render a <AddCustomerForm /> 1`] = `
<div>
  <form
    action="#"
    class="sc-bdfBQB IjRdz"
  >
    <div
      class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        aria-controls="panel1a-content"
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root"
        id="panel1a-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4"
          >
            Add a customer
          </h4>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="panel1a-header"
              id="panel1a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root sc-gKseQn etHLMD"
              >
                <div
                  class="sc-jSgvzq jqNVpG"
                >
                  <div
                    class="sc-eCstlR eLwwKz"
                  >
                    <b>
                      All the form fields are mandatory
                    </b>
                  </div>
                  <label
                    class="sc-gsTEea eRkGPM"
                    for="firstName"
                  >
                    First Name
                  </label>
                  <input
                    class="sc-dlfnuX eytFzo"
                    id="firstName"
                    name="firstName"
                    placeholder="John"
                    value=""
                  />
                  <label
                    class="sc-gsTEea eRkGPM"
                    for="lastName"
                  >
                    Last Name
                  </label>
                  <input
                    class="sc-dlfnuX eytFzo"
                    id="lastName"
                    name="lastName"
                    placeholder="Doe"
                    value=""
                  />
                  <label
                    class="sc-gsTEea eRkGPM"
                    for="phoneNumber"
                  >
                    Phone Number
                  </label>
                  <input
                    class="sc-dlfnuX eytFzo"
                    id="phoneNumber"
                    name="phoneNumber"
                    placeholder="+61 123 456 789"
                    type="tel"
                    value=""
                  />
                  <button
                    class="sc-hKgJUU boltyf"
                    data-testid="addCust-button"
                    type="submit"
                  >
                    Add Customer
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
